public/app/plugins/datasource/tempo/metricsSummary.ts

```diff
@@ -0,0 +1,294 @@
+import {
+  DataSourceInstanceSettings,
+  FieldDTO,
+  FieldType,
+  LoadingState,
+  MutableDataFrame,
+  sortDataFrame,
+} from '@grafana/data';
+
+export type MetricsSummary = {
+  spanCount: string;
+  errorSpanCount?: string;
+  p50: string;
+  p90: string;
+  p95: string;
+  p99: string;
+  series: Series[];
+};
+
+type Series = {
+  key: string;
+  value: {
+    type: number;
+    n?: number;
+    f?: number;
+    s?: string;
+    b?: string;
+    d?: string;
+    status?: number;
+    kind?: number;
+  };
+};
+
+type MetricsData = {
+  spanCount: string;
+  errorPercentage: string;
+  p50: string;
+  p90: string;
+  p95: string;
+  p99: string;
+  [key: string]: string | number;
+};
+
+export function createTableFrameFromMetricsSummaryQuery(
+  data: MetricsSummary[],
+  targetQuery: string,
+  instanceSettings: DataSourceInstanceSettings,
+  state: LoadingState
+) {
+  let frame;
+
+  if (state === LoadingState.Error) {
+    frame = emptyResponse;
+  } else if (state === LoadingState.Loading) {
+    frame = new MutableDataFrame({
+      name: 'Metrics Summary',
+      refId: 'metrics-summary',
+      fields: [
+        { name: 'state', type: FieldType.string, config: { displayNameFromDS: 'State', custom: { width: 300 } } },
+      ],
+      meta: {
+        preferredVisualisationType: 'table',
+      },
+    });
+    frame.add({
+      state: 'Loading...',
+    });
+  } else if (state === LoadingState.Done && data?.length) {
+    const dynamicMetrics: Record<string, FieldDTO> = {};
+    data.map((res: MetricsSummary) => {
```

Comment:
Please lets not use `map()` to do something else than actually map data. It's very confusing if map is use as an iterator that has side effects. Rather use `for` loop or `forEach()`.

Reply:
Ahh yes, good catch.